### PR TITLE
WIP: addRepoPermissions graphql api

### PR DESCRIPTION
**Pull request by mollylogue**
NOTE: This is currently a work in progress, but I'm opening this so that my teammates can take a look and leave feedback to ensure I'm going in the right direction. I'm still new to the permissions model.

Biggest known TODOs:
- Implement `AddRepoPendingPermissions` to add pending permissions
- Optimize queries

https://github.com/sourcegraph/sourcegraph/issues/28593

### CODE REVIEW

**Comment by indradhanush** [comment]
I took a first pass and this is all new to me. I will need to dig further to gather more context around how we're doing permissions management at the moment. In the meantime, if @unknwon has the capacity, he would be the best person to provide early feedback. I'll go ahead and request a review from him.

**Comment by unknwon** [comment]
Reviewing permissions store takes a lot metnal wrap up and energy, I'm in the middle of researching GitHub App, likely will do a full review next Monday (2021-12-13).

@mollylogue In case you missed it, GitHub has draft PR feature for WIP PRs.

CleanShot 2021-12-10 at 14 13 13@2x

**Comment by mollylogue** [comment]
@unknwon Sweet! I converted it to a draft for now. Feel free to wait until it's ready for review to take a look! Or if you want to look now, maybe focus on the sql queries because I have a feeling that's the biggest portion that I'll need feedback on.

File: enterprise/internal/database/perms_store.go [hunk 1]
```diff
@@ -1677,3 +1677,145 @@ func (s *PermsStore) observe(ctx context.Context, family, title string) (context
...
+	(%s, %s, %s, %s, %s)
+ON CONFLICT ON CONSTRAINT
+	repo_permissions_perm_unique
+DO UPDATE SET
+	user_ids_ints = CASE
+		WHEN excluded.user_ids_ints @> %s THEN
+			repo_permissions.user_ids_ints | excluded.user_ids_ints
+		ELSE
+			repo_permissions.user_ids_ints - %s::INT[]
+		END,
```
Comment (unknwon) [inline enterprise/internal/database/perms_store.go:1797]: This part looks quite different from the https://github.com/sourcegraph/sourcegraph/blob/721832c71bb3e127c962f3f50f986c81584523bd/enterprise/internal/database/perms_store.go#L835-L838

I'm curious what does `repo_permissions.user_ids_ints | excluded.user_ids_ints` give us?

**Comment by mollylogue** [comment]
Closing this for now until we gain more clarity around this explicit permissions work.

**Pull request closed without merging**